### PR TITLE
add support for custom outputs as part of the custom resource

### PR DESCRIPTION
**Pull request by nicka**
##### Status:

Ready

##### Description:

Essential for variable support in the future #1707.

##### Todos:

- [x] Write tests

### CODE REVIEW

File: lib/plugins/aws/deploy/lib/initializeResources.js [hunk 1]
```diff
@@ -49,6 +49,18 @@ module.exports = {
...
+        // check if the user has added some "custom Outputs"
+        if ('Outputs' in value) {
+          forEach(value.Outputs, (outputValue, outputKey) => {
+            const newOutputsObject = {
+              [outputKey]: outputValue,
+            };
+
+            merge(coreCFTemplate.Outputs, newOutputsObject);
+          });
+          delete value.Outputs; // eslint-disable-line no-param-reassign
```
Comment (nicka) [inline lib/plugins/aws/deploy/lib/initializeResources.js]: Not sure if this is preferable? But we really don't want those `Outputs` as part of the `Resource` 😊

File: lib/plugins/aws/deploy/README.md [hunk 2]
```diff
@@ -31,3 +31,23 @@ Furthermore it updates the stack with all the Resources which are defined in `se
 The stack status is checked every 5 seconds with the help of the CloudFormation API. It will return a success message if
 the stack status is `CREATE_COMPLETE` or `UPDATE_COMPLETE` (depends if you deploy your service for the first time or
 redeploy it after making some changes).
+
+### Custom outputs
+
```
Comment (pmuens) [inline lib/plugins/aws/deploy/README.md]: Could you write a short description what the purpose is?

**Comment by pmuens** [comment]
Thank you for that 🎉  👍 . Good job!

Added a minor comment about the docs.

@eahefnawy @flomotlik could you also take a look? Thanks!

File: lib/plugins/aws/deploy/README.md (hunk 2 above)
Comment (nicka) [inline lib/plugins/aws/deploy/README.md]: Sure! Coming later today 👍

**Comment by flomotlik** [comment]
Thanks @nicka for getting this started.

@pmuens @eahefnawy I'm currently thinking if instead of specifically adding the output section if we shouldn't make sure that everything that is in serverless.yml->resources simply gets merged into the CF template. So instead of handling resources and Output separately, we simply take every key that is in the resources section and simply merge them into our main template?

This way users could do anything they want in the custom resources.

**Comment by nicka** [comment]
@flomotlik No problem! This PR adds the option of adding outputs per custom resource, not one main Ouputs key. My reasoning was keeping outputs close to the custom resource. Less scrolling up and down(which happens a lot in big templates).

In `v1.0` we already has support for reffing an actual CF template file, which to me feels like better UX if you want to go big.

**Comment by pmuens** [comment]
@flomotlik @nicka 
Yes, merging everything inside the CloudFormation template would be a really flexible way to deal with the "`Outputs` problem" and makes Serverless even more powerful for AWS.

User are already familiar how that works and don't have to learn a new syntax.

**Comment by flomotlik** [comment]
@nicka just talked with @pmuens again and we decided that we want to stay 100% on cloud formation syntax in the resources section, so people don't have any surprises there.

Which means we can't merge this commit as it would include differences to cloud formation. I do think its an interesting feature though that could be nice as a separate plugin. So people could install and use the plugin to add output sections directly inside of resources, then the plugin could take them out of the resources and copy them into the actual output section (as you're doing).

If you're interested in implementing this as as plugin let me know, happy to help. This could then potentially be put into our community plugins repo discussed in #1722

**Comment by nicka** [comment]
>Which means we can't merge this commit as it would include differences to cloud formation.
@flomotlik @pmuens ☝️ I agree with you guys.

Now what's the way forward? Are you already working on a full merger of the CloudFormation template? I you need any help please let me know.

For now I will close this PR and I don't intend to work on a community plugin. As you already mention users should learn the actual CloudFormation syntax instead of mixing.

**Comment by flomotlik** [comment]
>Now what's the way forward? Are you already working on a full merger of the CloudFormation template? I you need any help please let me know.

This already works. Basically what you have in resources is a CloudFormation template. We're merging all the resources we create in there and use the resources object, stringify it and then send it to Cloudformation

**Comment by nicka** [comment]
@flomotlik So this already works with `Description`, `Outputs`, `Parameters` keys within the `resources` key?

**Comment by flomotlik** [comment]
@nicka Nope you're right my bad. Its still a plan but not yet actually implemented. It should be pretty simple and straight forward though. We're currently only reading them from `resources.Resources` in [initializeResources.js](https://github.com/serverless/serverless/blob/a525cf81968adc8a508eddf42247f3831c0a7be2/lib/plugins/aws/deploy/lib/initializeResources.js#L49) when we should actually just merge everything from `resources` into it. So basically in the iteration in there iterate over the outer resources element, not the inner one.

If you're interested in putting this in that would be great, otherwise I can open up another issue for it so it gets implemented in the future.

**Comment by nicka** [comment]
@flomotlik Sounds good to me! I should be able to have something up and running by Wednesday. Will create a PR when ready 👍

**Comment by nicka** [comment]
> Sounds good to me! I should be able to have something up and running by Wednesday. Will create a PR when ready 👍

@flomotlik @pmuens not sure if it's still relevant because of #1783?

**Comment by pmuens** [comment]
Yes, #1783 adds this feature so we should wait for the refactored implementation there. Would be great if you could review this code over there as well 👍

**Pull request closed without merging**